modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupReencryption.java

```diff
@@ -0,0 +1,583 @@
+import org.apache.ignite.internal.processors.cache.CacheGroupContext;
+import org.apache.ignite.internal.processors.cache.persistence.DbCheckpointListener;
+import org.apache.ignite.internal.processors.cache.persistence.GridCacheDatabaseSharedManager;
+import org.apache.ignite.internal.processors.cache.persistence.pagemem.PageMemoryEx;
+import org.apache.ignite.internal.util.future.GridCompoundFuture;
+import org.apache.ignite.internal.util.future.GridFinishedFuture;
+import org.apache.ignite.internal.util.future.GridFutureAdapter;
+import org.apache.ignite.internal.util.lang.IgniteInClosureX;
+import org.apache.ignite.internal.util.typedef.X;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.thread.IgniteThreadPoolExecutor;
+import org.apache.ignite.thread.OomExceptionHandler;
+
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_REENCRYPTION_BATCH_SIZE;
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_REENCRYPTION_DISABLED;
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_REENCRYPTION_THREAD_POOL_SIZE;
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_REENCRYPTION_THROTTLE;
+
+/**
+ * Cache group reencryption manager.
+ */
+public class CacheGroupReencryption implements DbCheckpointListener {
+    /** Thread prefix for reencryption tasks. */
+    private static final String REENCRYPT_THREAD_PREFIX = "reencrypt";
+
+    /** Max amount of pages that will be read into memory under checkpoint lock. */
+    private final int batchSize = IgniteSystemProperties.getInteger(IGNITE_REENCRYPTION_BATCH_SIZE, 1_000);
+
+    /** Timeout between batches. */
+    private final long timeoutBetweenBatches = IgniteSystemProperties.getLong(IGNITE_REENCRYPTION_THROTTLE, 0);
+
+    /** Disable background reencryption. */
+    private final boolean disabled = IgniteSystemProperties.getBoolean(IGNITE_REENCRYPTION_DISABLED, false);
+
+    /** Number of threads for partition scanning. */
+    private final int threadsCnt = IgniteSystemProperties.getInteger(IGNITE_REENCRYPTION_THREAD_POOL_SIZE,
+        Runtime.getRuntime().availableProcessors());
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** Lock. */
+    private final ReentrantLock lock = new ReentrantLock();
+
+    /** Mapping of cache group ID to reencryption context. */
+    private final Map<Integer, GroupReencryptionContext> grps = new ConcurrentHashMap<>();
+
+    /** Queue of groups waiting for a checkpoint. */
+    private final Queue<Integer> cpWaitGrps = new ConcurrentLinkedQueue<>();
+
+    /** Executor to start partition scan tasks. */
+    private final IgniteThreadPoolExecutor execSvc;
+
+    /** Stop flag. */
+    private boolean stopped;
+
+    /**
+     * @param ctx Grid kernal context.
+     */
+    public CacheGroupReencryption(GridKernalContext ctx) {
```

Comment:
Maybe we want to use current classes without context?  public CacheGroupReencryption(Logger log, etc?)

Reply:
Context is not fully initialized here, moreover I'm not sure that greatly increasing the number of arguments will improve the code. I understand that we want to loose the coupling, but I am not sure that it's easy to do here (class requires EncryptionManager, PageStoreManager, GridCacheDatabaseSharedManager and GridCacheProcessor at least).